Add a web scraping block

aickin opened the pull request:
This PR addresses #61, adding a `web_scrape` block to use browserless.io's web scraping API.

It takes in a URL and a CSS selector (which defaults to `body`), and it returns a JSON object that looks like:

```json
{
  "data": [
    {
      "selector": "body",
      "results": [
        {
          "html": <html of the selector>,
          "text": <text of the selector>,
          "attributes": [
          ]
        }
      ]
    },
  ],
  "response": {
    "code": <http status code from scraped site>,
    "status": <http status word from scraped site>,
    "url": <url from scraped site (including redirects)>,
    "ip": <ip from scraped site>,
    "port": <port from scraped site>,
  }
}
```

The `data` section might be unnecessarily complicated; it mirrors the raw response from Browserless, and I'm not sure if it's what we want or not. In particular, the Browserless API allows you to send an array of selectors, and `data` is an array of result sets, one for each selector sent in. Since we only allow one selector right now, it's a superfluous array in the result, and I don't know if we should keep it or not.

Everything is more or less copied over from the `search` block type, and I welcome feedback!

Pic:
<img width="476" alt="Screen Shot 2022-11-01 at 12 53 14 PM" src="https://user-images.githubusercontent.com/44199/199228750-9e4e8d2e-6e32-471e-a565-b2e9a4ac713c.png">

spolu commented:
In this very example where is the content? Can I get only the text?

aickin commented:
It's under `data[0].results[0]`. The attribute `html` has the html, and the attribute `text` has a text version of the page. I didn't expand in the pic because it's enormous.

spolu commented:
Ah I see it's under text :+1:

spolu commented on front/components/app/blocks/WebScrape.jsx:
```diff
@@ -0,0 +1,160 @@
+          </div>
+          <div className="flex w-full font-normal">
+            <div className="flex rounded-md flex-1">
+              Scrape this URL:&nbsp;
```
url

aickin replied:
Is there a reason HTML and CSS are capitalized but URL isn't? I don't care either way, just curious.

spolu replied:
Not really, we can go for URL :+1:

spolu commented on front/components/app/blocks/WebScrape.jsx:
```diff
@@ -0,0 +1,160 @@
+          </div>
+          <div className="flex flex-row ">
+            <div className="whitespace-nowrap font-normal">
+              Return the content from this CSS selector:&nbsp;
```
CSS selector

spolu commented on front/components/app/NewBlock.jsx:
```diff
@@ -33,6 +33,13 @@ export default function NewBlock({ spec, disabled, onClick, direction }) {
+      typeNames: ["web_scrape"],
+      name: "Web Scrape",
+      description:
+        "Download the html (or a portion of the html) of a page on the web.",
```
Download the HTML or text content of page on the web (or a portion of it)

spolu reviewed (commented):
How do you feel about renaming the block to `browser` ?

aickin commented on core/src/blocks/web_scrape.rs:
```diff
@@ -0,0 +1,149 @@
+
+
+#[async_trait]
+impl Block for Web_Scrape {
```
For the record, it's called `Web_Scrape` for consistency with `BlockType::Web_Scrape`, and that uses an underscord because the API fails if it's called `BlockType::WebScrape`. The way the serialization works, the block type names have to be the same as the lowercased version of the name of the block type in the spec. There may be a way to override the serialization to make this not necessary, but I didn't know enough about the API framework to figure it out.

In any case, it's moot if we change to a one word name like `Browser`, but it's something we'll run into again with two word block types.

aickin replied:
And `Browser`/`browser` sounds good to me!

spolu reviewed (changes requested):
Awesome! (re Browser name). Let's run with it then

aickin commented:
What do you think about the depth of the output JSON? Right now `data` is an array, but it will only ever have one element, because the block only allows one selector to be sent to browserless. I see three paths:

1. Leave the way it is in this PR.
2. Pull the first element of the data array out to be the value for `data`. This would make it easier to access `WEB_BLOCK.data.results[n].text` instead of `WEB_BLOCK.data[0].results[n].text`. However, if we decide to allow multiple CSS selectors in the future, we would have to change the shape of the JSON returned.
3. Add the ability to query multiple selectors to the block.

Thoughts?

spolu commented:
Leave it as is and we can add the ability to use multiple selectors later 👍

aickin commented:
OK, I think I addressed all of the code review comments in aaf11f5.

I also added a tiny feature in that commit that lops off "https://" from a URL at the beginning of an URL when you tab away from the url field. I was finding that I was pasting a lot of URLs from browser bars, and they always had the full scheme at the beginning, which is not desired behavior.
